webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/volume/VolumeSegmentIndexService.scala

```diff
@@ -0,0 +1,135 @@
+package com.scalableminds.webknossos.tracingstore.tracings.volume
+
+import com.google.inject.Inject
+import com.scalableminds.util.geometry.Vec3Int
+import com.scalableminds.util.tools.Fox
+import com.scalableminds.util.tools.Fox.box2Fox
+import com.scalableminds.webknossos.datastore.VolumeTracing.VolumeTracing
+import com.scalableminds.webknossos.datastore.VolumeTracing.VolumeTracing.{ElementClass => ElementClassProto}
+import com.scalableminds.webknossos.datastore.models.datasource.ElementClass
+import com.scalableminds.webknossos.datastore.geometry.{ListOfVec3IntProto, Vec3IntProto}
+import com.scalableminds.webknossos.datastore.helpers.ProtoGeometryImplicits
+import com.scalableminds.webknossos.datastore.models.{BucketPosition, UnsignedInteger, UnsignedIntegerArray}
+import com.scalableminds.webknossos.tracingstore.tracings.{FossilDBClient, KeyValueStoreImplicits, TracingDataStore}
+import com.typesafe.scalalogging.LazyLogging
+import net.liftweb.common.{Box, Empty, Failure, Full}
+import net.liftweb.util.Helpers.tryo
+
+import scala.concurrent.ExecutionContext
+
+// Segment-to-Bucket index for volume tracings in FossilDB
+// key: tracing id, segment id, mag – value: list of buckets
+// used for calculating segment statistics
+class VolumeSegmentIndexService @Inject()(val tracingDataStore: TracingDataStore)
+    extends KeyValueStoreImplicits
+    with ProtoGeometryImplicits
+    with VolumeBucketCompression
+    with SegmentIndexKeyHelper
+    with LazyLogging {
+
+  private val volumeSegmentIndexClient: FossilDBClient = tracingDataStore.volumeSegmentIndex
+
+  // Add segment index to merged tracing if all source tracings have a segment index
+  def shouldCreateSegmentIndexForMerged(tracings: Seq[VolumeTracing]): Boolean =
+    tracings.forall(_.hasSegmentIndex.getOrElse(false))
+
+  // Currently, segment index is not supported for volume tracings with fallback layer
+  def canHaveSegmentIndex(tracing: VolumeTracing): Option[Boolean] = Some(tracing.fallbackLayer.isEmpty)
+
+  def updateFromBucket(segmentIndexBuffer: VolumeSegmentIndexBuffer,
+                       bucketPosition: BucketPosition,
+                       bucketBytes: Array[Byte],
+                       previousBucketBytesBox: Box[Array[Byte]],
+                       elementClass: ElementClassProto)(implicit ec: ExecutionContext): Fox[Unit] =
+    for {
+      bucketBytesDecompressed <- tryo(
+        decompressIfNeeded(bucketBytes, expectedUncompressedBucketSizeFor(elementClass), "")).toFox
```

Comment:
Remove tryo/create new tryo fox function?

Reply:
Wrote a new issue https://github.com/scalableminds/webknossos/issues/7133 for this